### PR TITLE
Utilize persy index feature

### LINKED ISSUE #4: Fix multithread tests and stop ignoring

### LINKED ISSUE #6: Utilize persy index feature
This crate could probably utilize the persy index feature, which could reduce backend boilerplate code, althrough it would make the file format incompatible.

Partial solution for file format incompatibility:
* On (first?) ```LOAD```, check if file format is old or new, deserialize accordingly.
* On ```SAVE```, we need to know a preference (e.g. old or new), assume new, but provide a function to serialize to the old file format.

### PR DESCRIPTION
**Pull request by fogti**
This is a WIP to work on Issue #6.
It works, but requires a few changes:

- [X] waiting for [tglman/persy!6](https://gitlab.com/tglman/persy/merge_requests/6)
- [X] https://github.com/mount-tech/typedb/commit/9b09bb50d2cc3f181e5a7d8a46ea15d4f88728e5: we need to wait for the next persy release
- [X] https://github.com/mount-tech/typedb/commit/e01cbfd720e02d2eb714cb61ba30174e60a68a74: that commit should work fine, but it doesn't. It introduces race conditions, althrough the error probably isn't in that one (typedb) commit, something in the upstream (I tested against [zserik/persy@develop](https://gitlab.com/zserik/persy/tree/develop) and will soon verify against the master branch) might also be off...
- [X] squash resulting commits
- optional: check if caching (LRU?) is useful/needed, especially cache `key -> encode_kv(&key)` -> unneeded

cc @shockham @tglman

Closes #4, and closes #6.

### CODE REVIEW

**Comment by fogti** [comment]
Details about the second part-issue:
below is the backtrace of an infinitely hanging worker thread spawned by test `test_multithread_many_instance_insert`.

> 0  futex_reltimed_wait_cancelable (private=0, reltime=0x7f41e6ea7e30, expected=0, futex_word=0x7f41dc001508) at ../sysdeps/unix/sysv/linux/futex-internal.h:142
1  __pthread_cond_wait_common (abstime=0x7f41e6ea7f08, mutex=0x7f41e0001440, cond=0x7f41dc0014e0) at pthread_cond_wait.c:533
2  __pthread_cond_timedwait (cond=0x7f41dc0014e0, mutex=0x7f41e0001440, abstime=0x7f41e6ea7f08) at pthread_cond_wait.c:667
3  0x00005602f9427fc1 in std::sys::unix::condvar::Condvar::wait_timeout () at src/libstd/sys/unix/condvar.rs:102
4  0x00005602f936b33a in std::sys_common::condvar::Condvar::wait_timeout (self=0x7f41dc0014e0, mutex=0x7f41e0001440, dur=...) at /rustc/9af17757be1cc3f672928ecf06c40a662c5ec26d/src/libstd/sys_common/condvar.rs:51
5  0x00005602f935dda0 in std::sync::condvar::Condvar::wait_timeout (self=0x7f41dc000b30, guard=..., dur=...) at /rustc/9af17757be1cc3f672928ecf06c40a662c5ec26d/src/libstd/sync/condvar.rs:405
6  0x00005602f8cd161e in persy::index::config::Indexes::write_lock (self=0x7f41e0003058, indexes=...) at /home/zscheile/zseri/devel/persy/src/index/config.rs:459
7  0x00005602f73a1278 in persy::transaction::Transaction::prepare_commit (self=..., journal=0x7f41e0002df8, address=0x7f41e0002eb0, indexes=0x7f41e0003058, snapshots=0x7f41e00030b0, persy_impl=0x7f41e0002df0, allocator=0x7f41e00022e0) at /home/zscheile/zseri/devel/persy/src/transaction.rs:418
8  0x00005602f8cdd1cb in persy::persy::PersyImpl::prepare_commit (self=0x7f41e0002df0, tx=...) at /home/zscheile/zseri/devel/persy/src/persy.rs:482
9  0x00005602f8c8f1af in persy::Persy::prepare_commit (self=0x7f41e00038f0, tx=...) at /home/zscheile/zseri/devel/persy/src/lib.rs:1090
10 0x00005602f69a80cc in typedb::KV<K,V>::insert (self=0x7f41e00038f0, key=0x7f41e6ead704, value=0x7f41e6ead750) at /home/zscheile/zseri/devel/typedb/src/lib.rs:156
11 0x00005602f69ef8f5 in tests::test_multithread_many_instance_insert::{{closure}}::{{closure}} () at tests/tests.rs:304

**Comment by fogti** [comment]
ok, I verified it against the persy master branch. It hangs/stalls almost reliable.
The Travis job fails because of the `Cargo.toml` `[patch.crates-io]` section needed to use a custom persy source.

**Comment by tglman** [comment]
@zserik,

will checkout your project and test.

**Comment by tglman** [comment]
fixed in the last persy master, will release a 0.6.1 for backport the fix, the issue was in the unlock of the lock manager, the lock manager is based on wait conditions, I was unblocking only one thread and then thrown away the condition, making the other threads waiting on a "stale" condition, I made sure tu unblock all the threads.

**Comment by fogti** [comment]
@tglman
while running the currently ignored part of the test suite, I came across the following error:
```
thread '<unnamed>' panicked at 'assertion failed: !load.is_free()?', /home/zscheile/zseri/.cargo/git/checkouts/persy-39440952ebd0aaf2/47ab09c/src/allocator.rs:151:9
```
while persy is called via `persy.put::<ByteVec, ByteVec>` and `persy.prepare_commit`.

**Comment by fogti** [comment]
ok, I got a backtrace ([source code](https://github.com/zserik/persy/blob/fc1b68925e1659dc82471627710e57d3fb133bf4/tests/allocator_is_load.rs)):
```
     Running /home/zscheile/zseri/devel/persy/target/debug/deps/allocator_is_load-b480d0a47220836b

running 1 test
1009 9 9 1009 1019 19 19 1019 1029 29 29 1029 1039 39 39 1039 1049 49 49 1049 1059 59 59 1059 1069 69 69 1079 1069 79 79 1089 1079 89 89 1099 1089 99 99 1109 1099 test test_multithread_many_instance_insert ... test test_multithread_many_instance_insert has been running for over 60 seconds
109 109 1109 1119 119 119 1119 1129 129 129 1129 1139 139 139 1139 1149 149 149 1149 1159 1159 159 159 1169 1169 169 169 1179 179 179 1179 1189 1189 189 1199 189 1199 199 199 1209 1209 209 1219 209 219 219 1219 1229 229 229 1229 1239 239 239 1239 1249 249 249 1249 1259
thread '<unnamed>' panicked at 'assertion failed: !load.is_free()?', src/allocator.rs:151:9
stack backtrace:
   7: std::panicking::begin_panic
             at /rustc/9af17757be1cc3f672928ecf06c40a662c5ec26d/src/libstd/panicking.rs:411
   8: persy::allocator::Allocator::write_page
             at src/allocator.rs:151
   9: persy::allocator::Allocator::load_page
             at src/allocator.rs:145
  10: persy::journal::Journal::clear
             at src/journal.rs:184
  11: persy::transaction::Transaction::commit
             at src/transaction.rs:754
  12: persy::persy::PersyImpl::commit
             at src/persy.rs:515
  13: persy::Persy::commit
             at src/lib.rs:1139
  14: allocator_is_load::test_multithread_many_instance_insert::{{closure}}
             at tests/allocator_is_load.rs:23
thread '<unnamed>' panicked at 'persy prepare_commit: Lock', src/libcore/result.rs:1084:5
stack backtrace:
   8: rust_begin_unwind
             at src/libstd/panicking.rs:311
   9: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  10: core::result::unwrap_failed
             at src/libcore/result.rs:1084
  11: core::result::Result<T,E>::expect
             at /rustc/9af17757be1cc3f672928ecf06c40a662c5ec26d/src/libcore/result.rs:879
  12: allocator_is_load::test_multithread_many_instance_insert::{{closure}}
             at tests/allocator_is_load.rs:22
```

Code references:
https://github.com/zserik/persy/blob/fc1b68925e1659dc82471627710e57d3fb133bf4/src/allocator.rs#L143-L169
https://github.com/zserik/persy/blob/fc1b68925e1659dc82471627710e57d3fb133bf4/src/journal.rs#L184

**Comment by tglman** [comment]
@zserik,

Checkout trying to reproduce, it happen a bit randomly, (obviusly concurrency issue), hope to get on it soon

**Comment by fogti** [comment]
@shockham is the API change in this PR acceptable?

**Comment by tglman** [comment]
Hi @zserik,

Last persy master should fix this problem, it was an issue on the journal cleanup logic (simple mistake on a linked list), I tried multiple times without issues, but i will suggest you to try as well due to the random nature of this issue. will release another hotfix with this fix as well

**Comment by fogti** [comment]
@tglman yes, it works.

**Comment by fogti** [comment]
@tglman ok, I forked persy on gitlab.

File: Cargo.toml [hunk 1]
```diff
@@ -14,5 +14,7 @@ edition = "2018"
 bincode = "1.0"
 serde = "1.0"
 serde_derive = "1.0"
-log = "0.4"
-persy = "0.6"
+persy = "0.7"
+
+[patch.crates-io]
```
Comment (fogti) [inline Cargo.toml]: This section needs to be removed before merging.

**Comment by fogti** [comment]
ok, persy PR !11 is merged, I reverted the Cargo.toml change.

**Comment by tglman** [comment]
Hi, 
Persy 0.7.x is out with all require changes, fee free to merge and push ahead

**Comment by shockham** [comment]
> @shockham is the API change in this PR acceptable?

Sorry hadn't responded on this earlier, I am not a huge fan of this wholesale change to this project. Some of the changes come down to personal preference, and don't necessarily align with mine which is to keep this as simple and minimal as possible.

But feel free to create a new forked project out of the code in this PR, as it basically replaces the whole of `src/lib.rs` anyway!

**Comment by fogti** [comment]
ok, I will try to rebase it against the new master and drop the `patch` section from Cargo.toml. I'll probably create my own fork if that's the way to go.

**Review by fogti — COMMENTED** [review]
fixed.

**Comment by fogti** [comment]
closed with reason as stated above: https://github.com/mount-tech/typedb/pull/8#issuecomment-542648669

**Comment by fogti** [comment]
I'll probably claim the crate name [`perkv`](https://github.com/zserik/perkv) for the fork.

**Pull request closed without merging**